src/operator/tensor/elemwise_binary_op.h

```diff
@@ -25,363 +25,938 @@
+                                 mshadow::Tensor<xpu, 2, DType> *out,
+                                 const size_t iter_out) {
+    using namespace mxnet_op;
+    using namespace mshadow::expr;
+    const int index_out_min = std::min(idx_l, idx_r);
+    if (index_out_min > iter_out) {
+      const size_t size = (*out)[iter_out].shape_.Size();
+      const DType zero_input_val = OP::Map(DType(0), DType(0));
+      #pragma omp parallel for
+      for (int i = static_cast<int>(iter_out); i < index_out_min; ++i) {
+        MXNET_ASSIGN_REQ_SWITCH(req, Req, {
+          Kernel<MapSetToScalar<Req>, xpu>::Launch(s, size, (*out)[i].dptr_, zero_input_val);
+        });
+      }
+    }
+    return static_cast<size_t>(index_out_min);
   }
-  if (std::is_same<ROP, mshadow_op::identity>::value && req[1] == kWriteInplace) {
-    CHECK_EQ(ograd_dptr, rgrad_dptr);
-  } else if (req[1] != kNullOp) {
-    MXNET_ASSIGN_REQ_SWITCH(req[1], Req,
-      {Kernel<BinaryOpBackwardUseNone<ROP, Req>, xpu>::Launch(s, size, rgrad_dptr,
-        ograd_dptr);});
+
+  template<typename DType>
+  static inline bool IsSameArray(const NDArray *a1, const NDArray *a2) {
+    if (a1 && a2) {
+      if (a1 == a2) {
+        return true;
+      }
+      if (a1->ctx().dev_type == a2->ctx().dev_type && a1->ctx().dev_id == a2->ctx().dev_id) {
+        const DType *pa1 = a1->data().dptr<DType>();
+        if (pa1 && pa1 == a2->data().dptr<DType>()) {
+          return true;
+        }
+      }
+    }
+    return false;
   }
-}
-
-// TODO(haibin) This is a single-thread inefficient implementation
-// This implementation only works on CPU
-template<typename xpu, typename OP>
-void BinaryComputeRspRspImpl(const nnvm::NodeAttrs& attrs,
-                             const OpContext& ctx,
-                             const std::vector<NDArray>& inputs,
-                             const std::vector<OpReqType>& req,
-                             const std::vector<NDArray>& outputs) {
-  if (req[0] == kNullOp) return;
-  CHECK(req[0] == kWriteTo) << "only kWriteTo is supported for rowsparse elemwise_add";
-  using namespace rowsparse;
-  using namespace mshadow;
-  auto &lhs = inputs[0];
-  auto &rhs = inputs[1];
-  auto &output = outputs[0];
-
-  bool init_l = lhs.storage_initialized();
-  bool init_r = rhs.storage_initialized();
-  Stream<xpu> *s = ctx.get_stream<xpu>();
-  // both inputs are zeros
-  if (!init_l && !init_r) {
-    NDArray out = output;
-    FillZerosRspImpl(s, &out);
-    return;
+
+  /*! \brief For some types or sparse, we can assume 100% density can be computer
+   * more quickly with the standard dense procedure
+   * @param arr the array to test
+   * @return bool, whether the array is effectively dense
+   */
+  static inline bool IsEffectivelyDense(const NDArray& arr) {
+    switch (arr.storage_type()) {
+      case kDefaultStorage:
+        return true;
+      case kRowSparseStorage:
+        return arr.shape().Size() == arr.aux_shape(rowsparse::kIdx).Size();
```

Comment:
Historically, checking storage_shape() was not a sufficient check (recall the bug).  At any rate, this yields a correct result.